### PR TITLE
USA Scraper Cache Fix

### PR DESCRIPTION
**Pull request by DGieseke**
This PR amends the cache logic within the `Scraper` class such that action dates are normalized at the time of deciding whether or not to send bill data for processing.  Previously, the USA scraper was scraping dates in a format that was bypassing our cache--this ensures we aren't reprocessing old bills at time of scrape.

New Function: `normalize_action_dates`

Testing: This was tested with our USA scraper module as well as our CA and NC scraper modules for good measure.

### CODE REVIEW

**Review by rob-abstract — DISMISSED** [review]

File: openstates/scrape/base.py [hunk 1]
```diff
@@ -442,11 +442,30 @@ def save_object(self, obj):
                         new_json = replace_none_in_dict(new_json)
                         existing_json = replace_none_in_dict(existing_json)
 
-                        mismatched_fields = {
-                            key
-                            for key in new_json.keys()
-                            if new_json[key] != existing_json.get(key)
-                        }
+                        def normalize_action_dates(actions):
```
Comment (alanmf) [inline openstates/scrape/base.py:445]: do we maintain tests of our own here? seems like something that wants to be tested.

File: openstates/scrape/base.py [hunk 2]
```diff
@@ -442,11 +442,30 @@ def save_object(self, obj):
...
+                            """Normalize action date fields to just the date part for comparison."""
+                            normed = []
+                            for action in actions:
+                                action = dict(action)
+                                if "date" in action and action["date"]:
+                                    # Only keep YYYY-MM-DD part
+                                    action["date"] = action["date"][:10]
```
Comment (alanmf) [inline openstates/scrape/base.py]: this is scary, perhaps a call to some strip/trim for whitespace at least could be helpful? what happens if action["date"] is null? [:10] would fail right? so perhaps a check it is at least that long?

Comment (DGieseke) [reply openstates/scrape/base.py]: Thanks for flagging this! I made the changes:

1. **Tests** – Added coverage in `test_scraper.py` for date normalization edge cases, bill comparison logic, and crash scenarios.  
2. **Date normalization** – Replaced the unsafe `[:10]` slice with a safer approach:
   - Trim whitespace
   - Check string length ≥10
   - Validate `YYYY-MM-DD` format with dashes
   - Handle null values gracefully
   - Log a warning instead of crashing if the format is unexpected  
3. **Bonus** – Fixed a duplicate `json.dump` call that was causing test failures.

`action['date']` is typed as a date in the OS schema, but these safeguards should help with edge cases.

File: openstates/scrape/base.py [hunk 3]
```diff
@@ -442,11 +442,35 @@ def save_object(self, obj):
...
-                        }
+                        def normalize_action_dates(actions):
+                            """Normalize action date fields to just the date part for comparison."""
+                            normed = []
+                            for action in actions:
+                                action = dict(action)
+                                if "date" in action and action["date"]:
+                                    date_str = str(action["date"]).strip()
+                                    # Only slice if we have at least 10 characters and it looks like a date
+                                    if len(date_str) >= 10 and date_str[4] == '-' and date_str[7] == '-':
```
Comment (alanmf) [inline openstates/scrape/base.py]: you should be able to use the datetime lib here instead of looking at the string indices
something like this:
        try:
            parsed = datetime.strptime(date_str[:10], "%Y-%m-%d")
            action["date"] = parsed.date().isoformat()

Comment (DGieseke) [reply openstates/scrape/base.py]: Revised and retested! Thanks for the advice here.

**Review by alanmf — APPROVED** [review]
